### ISSUE TITLE
edgeX message bus sink not working

### ISSUE THREAD

**Issue by Putitii** [lf-edge/ekuiper#831]
Im doing a project on EdgeX for my university where I read temperature/humidity values from a sensor (my case it's a python program) and send them to Azure cloud. I can export the values to Azure from the event topic normally without issue. The issue begins when I try to use the Kuiper Engine to do some edge computing since I can't seem to get it to output the analysis results to the message hub.

I created this stream:
![image](https://user-images.githubusercontent.com/66571416/121687057-cd79bc80-cab9-11eb-9b58-27b0d4bdfbfd.png)

Then I created this test rule:
![image](https://user-images.githubusercontent.com/66571416/121687385-33feda80-caba-11eb-91d1-e2e36d57c07a.png)

When I go to the container logs: (docker logs edgex-kuiper):
![image](https://user-images.githubusercontent.com/66571416/121687756-9c4dbc00-caba-11eb-96f0-db1b6803f896.png)
Is it suppose to print some message about publishing to the new message bus?

In the Azure-Export-Service (https://github.com/edgexfoundry/edgex-examples/tree/master/application-services/custom/azure-export) I used this configuration for the new message bus:
![image](https://user-images.githubusercontent.com/66571416/121688389-4c232980-cabb-11eb-8d25-4ba8a3c48a40.png)

In the export-service it just stays waiting for something in the application topic, which never appears:
![image](https://user-images.githubusercontent.com/66571416/121688597-8bea1100-cabb-11eb-9a55-dbbc33d7b8a7.png)

How do I access the new values sent to the message bus? I don't understand why it's not working, since I followed the steps in the documentation (https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/rules/sinks/edgex.md)

Thank you for the help.

**Comment by ngjaying** [comment]
@Putitii Thanks for the very detail explanation of the problem. It is very helpful to understand your problem. 

> Is it suppose to print some message about publishing to the new message bus?
No in normal mode, only when debug mode is open (edit kuiper.yaml, enable debug and restart). Since you also add a sink node and get the result, it indicates the rule is running.  And the result is empty "[{}]" which would be correct if you did not send temperature and humidity reading from your source. If edgex sink is working properly, you should also receive an event with empty readings. I see you are verifying that with Azure-Export-Service.  Unfortunately, I am not very familiar with the Azure-Export-Service. I think you can first verify it by other ways. For example, create a zero mq client to subscribe for edgex message bus and see if you receive the data. Check this [sub example](https://github.com/lf-edge/ekuiper/blob/master/fvt_scripts/edgex/sub/sub.go).

Additionally, [eKuiper supports to integrate with Azure IOT hub directly](https://emqx.medium.com/lightweight-edge-computing-emq-x-kuiper-and-azure-iot-hub-integration-solution-151134ead024). You may be interesting to have a try which will bypass the edgex sink.

**Comment by Putitii** [comment]
hello, I've been trying get the edgex sink to work and I figured how to enable DEBUG mode.
![image](https://user-images.githubusercontent.com/66571416/122545581-e9cfa900-d025-11eb-9c32-f10afa71a212.png)

It says that it's publishing, so how do I see the information in the port 5571?

I also used the Azure IoT Hub MQTT sink and it still doesn't work (it says network error):
The rule:
![image](https://user-images.githubusercontent.com/66571416/122550838-14246500-d02c-11eb-9261-171a7650b29b.png)
The edgex-kuiper container output:
![image](https://user-images.githubusercontent.com/66571416/122551357-c5c39600-d02c-11eb-9575-59b867c51bd8.png)

I also tried with "sub example" that you suggested but I get this (added prints for debug):
Added this to the line 75: 
![image](https://user-images.githubusercontent.com/66571416/122551758-55694480-d02d-11eb-8163-f70ca9fd0141.png)

 and executed the program:
![image](https://user-images.githubusercontent.com/66571416/122551680-39fe3980-d02d-11eb-8587-64098852970c.png)

**Comment by ngjaying** [comment]
> It says that it's publishing, so how do I see the information in the port 5571?

It is in edgex message bus, the type is zmq by default. It is a edgex question for how to read the data. But it is basically to create a zmq client and subscribe to the events. We use the sub program to verify the result. But please use `subEventsFromZMQ`. 

Your using of sub program is weird. Line 75 is actually the error branch for connection, why did you add debug there? You can just uncomment the log.Errorf to print out the error message when error happens. And remember to use **subEventsFromZMQ** instead.



> I also used the Azure IoT Hub MQTT sink and it still doesn't work (it says network error)

I think your mqtt connection info is invalid. Can you try in mqtt client (mosquitto or [mqttx](https://github.com/emqx/MQTTX)) directly to verify the connection info?

**Comment by Putitii** [comment]
Hello again, thank you so much for the help!

I got it working fine. I did all your suggestions and none worked. 
I always thought my main problems were with the certificates, and which one to choose, etc.
So I tried to create a new device but , instead of using Iot Hub with X.509 CA certificate, I selected the "Simmetric Key", and then I generated a SAS token.

Then I did the MQTT IotHub Sink that you suggested and it worked with the rule in the next picture. You were right, my connection info was incorrect. I configued a mosquitto mqtt client correctly and got it working and then used the same info on the rule: 

![image](https://user-images.githubusercontent.com/66571416/122988575-cd22d080-d399-11eb-9bfe-1d6297aa43bf.png)
Note: I was getting empty readings in Iot Hub, then I used the "omitIfEmpty: true" and then it stopped sending empty readings from Kuiper.

Got the readings in VS Code (with Azure IoT Tools):

![image](https://user-images.githubusercontent.com/66571416/122988719-f3e10700-d399-11eb-9022-b478a368487c.png)

Thank you so much for the help!
Hopefully this helped someone that has the same issue.
